wrapper/src/main/java/software/amazon/jdbc/util/monitoring/MonitorServiceImpl.java

```diff
@@ -0,0 +1,232 @@
+    if (isInitialized.get()) {
+      return;
+    }
+
+    initLock.lock();
+    try {
+      if (isInitialized.get()) {
+        return;
+      }
+
+      cleanupExecutor.scheduleAtFixedRate(
+          this::checkMonitors, cleanupIntervalNanos, cleanupIntervalNanos, TimeUnit.NANOSECONDS);
+      cleanupExecutor.shutdown();
+      isInitialized.set(true);
+    } finally {
+      initLock.unlock();
+    }
+  }
+
+  protected void checkMonitors() {
+    LOGGER.finest(Messages.get("MonitorServiceImpl.checkingMonitors"));
+    for (ExpirationCache<Object, MonitorItem> cache : monitorCaches.values()) {
+      // Note: the map returned by getEntries is a copy of the ExpirationCache map
+      for (Map.Entry<Object, MonitorItem> entry : cache.getEntries().entrySet()) {
+        MonitorItem monitorItem = entry.getValue();
+        if (monitorItem == null) {
+          continue;
+        }
+
+        Monitor monitor = monitorItem.getMonitor();
+        if (monitor == null) {
+          continue;
+        }
+
+        if (monitor.getState() == MonitorState.STOPPED) {
+          cache.remove(entry.getKey());
+        }
+
+        MonitorSettings monitorSettings = monitorSettingsByType.get(monitor.getClass());
+        if (monitorSettings == null) {
+          cache.removeIfExpired(entry.getKey());
+          continue;
+        }
+
+        Set<MonitorErrorResponse> errorResponses = monitorSettings.getErrorResponses();
+        if (System.nanoTime() - monitor.getLastUsedTimestampNanos() < monitorSettings.getInactiveTimeoutNanos()) {
+          // Monitor has updated its last-used timestamp recently and is not considered stuck.
+          cache.removeIfExpired(entry.getKey());
+        }
+
+        // Monitor has been inactive for longer than the inactive timeout and is considered stuck.
+        LOGGER.fine(
+            Messages.get("MonitorServiceImpl.monitorStuck",
+                new Object[]{monitor, TimeUnit.NANOSECONDS.toSeconds(monitorSettings.getInactiveTimeoutNanos())}));
+        if (errorResponses.contains(MonitorErrorResponse.RESTART)) {
+          // Note: the put method disposes of the old item
+          LOGGER.fine(Messages.get("MonitorServiceImpl.restartingMonitor", new Object[]{monitor}));
+          cache.put(entry.getKey(), new MonitorItem(monitorItem.getMonitorSupplier()));
+        } else {
+          cache.remove(entry.getKey());
+        }
+      }
+    }
+  }
+
+  @Override
+  public <T extends Monitor> void registerMonitorTypeIfAbsent(
+      Class<T> monitorClass,
+      long timeToLiveNanos,
+      long inactiveTimeoutNanos,
+      Set<MonitorErrorResponse> errorResponses,
+      @Nullable ShouldDisposeFunc<T> shouldDisposeFunc) {
+    monitorCaches.computeIfAbsent(
+        monitorClass,
+        mc -> {
+          monitorSettingsByType.putIfAbsent(monitorClass, new MonitorSettings(inactiveTimeoutNanos, errorResponses));
+          ShouldDisposeFunc<MonitorItem> wrappedShouldDisposeFunc = shouldDisposeFunc == null ? null
```

Comment:
I created the MonitorItem inner class so that monitors can be stored together with the suppliers used to create them. This allows us to re-create the monitor if it encounters an error or gets stuck. However, it makes things messy in this function. The caller passes in a shouldDisposeFunc for the monitor itself (eg ClusterTopologyMonitorImpl), but since we are actually storing MonitorItem's in the cache, we need to pass ShouldDisposeFunc\<MonitorItem>. For now I resolved this by wrapping the passed in lambda, but this seems messy and I am getting a warning when I try to cast monitorItem.getMonitor() to T.